### PR TITLE
feat: message silver and core table

### PR DESCRIPTION
**Pull request by geewynn**
# Description
Message core__fact and silver__message table

# Tests

- [x] Please provide evidence of your successful `dbt run` / `dbt test` here
- [x] Any comparison between `prod` and `dev` for any schema change
image
image
image
image

# Checklist
- [x] Follow [dbt style guide](https://github.com/dbt-labs/corp/blob/main/dbt_style_guide.md)
- [x] Tag the person(s) responsible for reviewing proposed changes
- [x] Notes to deployment, if a `full-refresh` is needed for any table
- [x] Run `git merge main` to pull any changes from remote into your branch prior to merge.

### CODE REVIEW

File: models/silver/silver__messages.sql [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+            final_attrib_obj
+        ) over (
+            PARTITION BY tx_id,
+            msg_index
+        ) AS attributes,
+        _ingested_at,
+        _inserted_timestamp
+    FROM
+        final_table
+    ORDER BY
```
Comment (eevui) [inline models/silver/silver__messages.sql:210]: I think we should leave ordering out for the analyst by default.

File: models/silver/silver__messages.sql [hunk 2]
```diff
@@ -0,0 +1,217 @@
+{{ config(
+    materialized = 'incremental',
+    cluster_by = ['_inserted_timestamp::DATE'],
+    unique_key = 'tx_id'
+) }}
+
+WITH txs AS (
```
Comment (eevui) [inline models/silver/silver__messages.sql:7]: Follow dbt guidelines: no caps for sql keywords.

File: models/silver/silver__messages.sql [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        msg_index,
+        msg_type,
+        INDEX AS attribute_index,
+        VALUE AS ATTRIBUTE,
+        VALUE :key :: STRING AS attribute_key,
+        IFF(
+            VALUE :key = 'amount',
+            SPLIT_PART(
+                TRIM(
+                    REGEXP_REPLACE(
```
Comment (eevui) [inline models/silver/silver__messages.sql:74]: I take this regexp selects the digit portion? Consider `regexp_substr` to select a substring

File: models/silver/silver__messages.sql [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+                attribute_key
+        ) - 1 AS key_index,
+        _ingested_at,
+        _inserted_timestamp
+    FROM
+        events,
+        LATERAL FLATTEN(
+            input => event_attributes
+        )
+    ORDER BY
```
Comment (eevui) [inline models/silver/silver__messages.sql:120]: Is ordering here important?

File: models/silver/silver__messages.yml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      - name: attributes
+        description: "{{ doc('attributes') }}"
+        tests:
+          - not_null
+          - dbt_expectations.expect_column_values_to_be_in_type_list:
+              column_type_list:
+                - OBJECT
+
+      - name: _ingested_at
+        description: "{{ doc('_ingested_at') }}"
```
Comment (eevui) [inline models/silver/silver__messages.yml:84]: Lacking tests.

File: models/core/core__fact_messages.sql [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    block_id,
+    tx_id,
+    tx_succeeded,
+    chain_id,
+    message_index,
+    message_type,
+    attributes
+FROM
+    messages
+
```
Comment (eevui) [inline models/core/core__fact_messages.sql:25]: Please follow style guide: no caps on keywords.

Also, perhaps use a `final` CTE.

File: models/silver/silver__messages.sql [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        block_timestamp,
+        block_id,
+        tx,
+        tx_succeeded,
+        VALUE :events AS logs,
+        VALUE :msg_index :: NUMBER AS msg_index,
+        tx :body :messages [0] :"@type" :: STRING AS msg_type,
+        _ingested_at,
+        _inserted_timestamp
+    FROM
```
Comment (eevui) [inline models/silver/silver__messages.sql:20]: Missing incremental filter.

File: models/silver/silver__messages.sql [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        LATERAL FLATTEN(
+            input => tx :tx_result :log
+        )
+),
+blocks AS (
+    SELECT
+        block_id,
+        chain_id
+    FROM
+        {{ ref('silver__blocks') }}
```
Comment (eevui) [inline models/silver/silver__messages.sql:31]: Missing incremental filter _or_ cut down on this `blocks` CTE ASAP. Take only rows that you need.

File: models/silver/silver__messages.sql [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        block_id,
+        final_attrib_obj,
+        _ingested_at,
+        _inserted_timestamp
+    FROM
+        third_table
+),
+FINAL AS (
+    SELECT
+        DISTINCT CONCAT(
```
Comment (eevui) [inline models/silver/silver__messages.sql:187]: Don't think we need a `distinct` here any more?

File: models/core/core__fact_messages.yml [hunk 10]
```diff
@@ -0,0 +1,81 @@
+version: 2
+
+models:
+  - name: core__fact_messages
+    description: |-
+      This table records all the transactions of the Terra2 blockchain.
+
+    columns:
+      - name: message_id
```
Comment (eevui) [inline models/core/core__fact_messages.yml:9]: Add `unique` test to primary keys.

**Review by eevui — CHANGES_REQUESTED** [review]
Got a couple comments.

For all sql: no caps on keywords.

File: models/silver/silver__messages.sql (hunk 7 above)
Comment (D0R11) [reply models/silver/silver__messages.sql:20]: agree

File: models/silver/silver__messages.sql (hunk 8 above)
Comment (D0R11) [reply models/silver/silver__messages.sql:31]: chain id is needed according to what was agreed upon in discord but it's only in the `blocks` table so I had to fetch it there

File: models/silver/silver__messages.sql (hunk 9 above)
Comment (D0R11) [reply models/silver/silver__messages.sql:187]: Running it without distinct here would make an error duplicate key. You can try the query [here](https://app.snowflake.com/us-east-1/dqa62717/w4X7vAXtGbri#query).

File: models/core/core__fact_messages.yml (hunk 10 above)
Comment (D0R11) [reply models/core/core__fact_messages.yml:9]: agree

File: models/silver/silver__messages.sql (hunk 2 above)
Comment (D0R11) [reply models/silver/silver__messages.sql:7]: `::DATE` is not included in the column name too, and the unique id is the newly-created `message_id` instead

File: models/silver/silver__messages.sql (hunk 4 above)
Comment (D0R11) [reply models/silver/silver__messages.sql:120]: Ah, no. Should have removed that, apologies

File: models/silver/silver__messages.sql (hunk 3 above)
Comment (D0R11) [reply models/silver/silver__messages.sql:74]: yup, was originally trying to do something else so I tried web3_analyst's suggestion but you're right. The code could have been shorter, sorry

File: models/silver/silver__messages.sql (hunk 8 above)
Comment (eevui) [reply models/silver/silver__messages.sql]: Ah yeah, either join with transactions or perform incremental filter. Blocks and its txs are ingested at the same time, afaik.

Comment (D0R11) [reply models/silver/silver__messages.sql]: I added an incremental filter on the latest commit can you check?

File: models/core/core__fact_messages.yml (hunk 10 above)
Comment (D0R11) [reply models/core/core__fact_messages.yml]: done

File: models/core/core__fact_messages.sql (hunk 6 above)
Comment (D0R11) [reply models/core/core__fact_messages.sql:25]: I think this is okay

**Review by eevui — APPROVED** [review]
LGTM!

**Pull request merged**